File: chat/simple/chat_client.py
import sys
import socket

BUF_SIZE = 1024

if __name__ == "__main__":

  if len(sys.argv) != 3:
    sys.stderr.write("usage: chat_client ip port\n")
    exit(1)

  try:
    addr = sys.argv[1]
    port = int(sys.argv[2])
    assert port > 0 
  except:
    sys.stderr.write("error: invalid port\n")
    exit(1)

  sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
  try:
    sock.connect((addr, port))
  
    print 'Connected...'
    print 'Send and receive messages. Empty line closes connection.'

    while True:
      data = raw_input('Msg:')
      if data == '':
        break
      sock.send(data)
      data = sock.recv(BUF_SIZE)
      if not data:
        break
      print 'Answer:', data
    sock.close()

  except socket.error, e:
    print 'Error:', e

File: chat/simple/dotnet/ChatClient.cs
using System;
using System.Linq;
using System.Net;
using System.Net.Sockets;
using static System.Text.Encoding;

namespace SocketChatClient
{
    class Program
    {
        private static readonly int BUFFER_SIZE = 1024;
        private static IPAddress _address;
        static void Main(string[] args)
        {
            if (args.Length != 2)
            {
                Console.Error.WriteLine("Usage: run program only with parameters <ip> <port>");
                Environment.Exit(1);
            }

            try
            {
                _address = IPAddress.Parse(args[0]);
            }
            catch (Exception e)
            {
                Console.Error.WriteLine("Invalid IP address");
                Environment.Exit(1);
            }
            bool portIsValid = int.TryParse(args[1], out int port);
            if (!portIsValid || port <= 0)
            {
                Console.WriteLine("Invalid port number provided");
                Environment.Exit(1);
            }
            var endPoint = new IPEndPoint(_address, port);
            using (var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
            {
                try
                {
                    socket.Connect(endPoint);
                }
                catch (Exception e)
                {
                    Console.Error.WriteLine($"Couldn't connect to {endPoint.Address}:{endPoint.Port}");
                    Environment.Exit(1);
                }
                Console.WriteLine("Enter message to send, enter whitespace to exit");
                while (true)
                {
                    Console.Write("Enter your message: ");
                    var message = Console.ReadLine();
                    if (string.IsNullOrWhiteSpace(message)) { break; }
                    var data = ASCII.GetBytes(message);
                    socket.Send(data);
                    Console.WriteLine("Message Sent.");

                    var buffer = new byte[BUFFER_SIZE];
                    int dataSize = socket.Receive(buffer);
                    var responseMsg = ASCII.GetString(buffer
                        .Take(dataSize)
                        .ToArray());
                    if (string.IsNullOrWhiteSpace(responseMsg)) { break; }
                    Console.WriteLine($"Message received: {responseMsg}");
                }
            }
        }
    }
}
